apps/docs/pages/guides/auth/server-side/nextjs.mdx

```diff
@@ -0,0 +1,947 @@
+
+  </div>
+  <div className="border-b mt-3 pb-3">
+    <Accordion.Item
+      header={<span className="text-foreground">What does the `cookies` option do?</span>}
+      id="client-storage-cookies"
+    >
+
+    The cookies option lets the Supabase client know how to access the cookies, so it can read and write the user session data. To make `@supabase/ssr` framework-agnostic, the cookies methods aren't hard-coded. But you only need to set them up once. You can then reuse your utility functions whenever you need a Supabase client.
+
+    </Accordion.Item>
+
+  </div>
+</Accordion>
+
+</StepHikeCompact.Details>
+
+<StepHikeCompact.Code>
+
+<CH.Code className="min-h-[34rem]">
+
+```ts utils/supabase/client.ts
+import { createBrowserClient } from '@supabase/ssr'
+
+export function createClient() {
+  return createBrowserClient(
+    process.env.NEXT_PUBLIC_SUPABASE_URL!,
+    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!
+  )
+}
+```
+
+```ts utils/supabase/server.ts
+import { createServerClient } from '@supabase/ssr'
+import { cookies } from 'next/headers'
+
+export function createClient() {
+  const cookieStore = cookies()
+
+  return createServerClient(
+    process.env.NEXT_PUBLIC_SUPABASE_URL!,
+    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
+    {
+      cookies: {
+        get(name: string) {
+          return cookieStore.get(name)?.value
+        },
+      },
+    }
+  )
+}
+```
+
+```ts utils/supabase/actions.ts
+import { type CookieOptions, createServerClient } from '@supabase/ssr'
+import { cookies } from 'next/headers'
+
+export function createClient() {
+  const cookieStore = cookies()
+
+  return createServerClient(
+    process.env.NEXT_PUBLIC_SUPABASE_URL!,
+    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
+    {
+      cookies: {
+        get(name: string) {
+          return cookieStore.get(name)?.value
+        },
+        set(name: string, value: string, options: CookieOptions) {
+          return cookieStore.set({ name, value, ...options })
+        },
+        remove(name: string, options: CookieOptions) {
+          return cookieStore.set({ name, value: '', ...options })
+        },
+      },
+    }
+  )
+}
+```
+
+</CH.Code>
+
+</StepHikeCompact.Code>
+
+</StepHikeCompact.Step>
+
+<StepHikeCompact.Step step={4}>
+
+<StepHikeCompact.Details title="Hook up middleware">
+
+Create a `middleware.ts` file at the root of your project.
+
+Since Server Components can't write cookies, you need middleware to refresh expired Auth tokens and store them. Copy the code to refresh the Auth token whenever it's expired.
+
+Add a matcher so the middleware doesn't run on routes that don't access Supabase.
+
+<Admonition type="danger">
+
+The middleware uses `supabase.auth.getSession()`.
+
+For performance, this function doesn't revalidate the user's Auth token except when the cookie's token expires.
+
+If you extend your middleware to process user data, replace `getSession` with `getUser` to ensure that user data is always revalidated.
```

Comment:
We need to be much clearer and more paranoid in wording about this, as folks won't get how serious it can be. Especially since Next.js and this whole SSR is brand new and people are not used to thinking about cookies as they've been out of fashion for 15 years now.

```suggestion
Be careful when protecting pages. Your user's session is stored in Cookies, which means that anyone can spoof values in this header. Avoid using `supabase.auth.getSession()` to protect pages and instead use `supabase.auth.getUser()`.

Do not blindly trust the return values of `supabase.auth.getSession()` inside server components, as the returned object could be controlled by an attacker. It's safe to trust `supabase.auth.getUser()` because it sends a request to Supabase Auth which validates the access token.
```

Reply:
Thanks, good point. I've reworded it a little 👇🏼 , then added it to the Server Component step as well, and changed the middleware to use `getUser` just in case

```
Be careful when protecting pages. The server gets the user session from the cookies, which can be spoofed by anyone.

Always use `supabase.auth.getUser()` to protect pages and user data.

_Never_ trust `supabase.auth.getSession()` inside Server Components. It isn't guaranteed to revalidate the Auth token.

It's safe to trust `getUser()` because it sends a request to the Supabase Auth server every time to revalidate the Auth token.
```

Reply:
this section was great btw — I did not know this